### DISCUSSION TITLE
TriggerIndexer Delayed Loading At Runtime

### DISCUSSION THREAD

**Discussion by mickdelaney** [elsa-workflows/elsa-core#1603] [General]
Hi,

I'm seeing that sometimes the TriggerStore has not loaded Triggers before I execute a _WorkflowLaunchpad.ExecutePendingWorkflowsAsync_

This happens where I have a service starting up, and it then starts processing Kafka messages, but the presumably the IStartupTask hasnt completed. 

I've tried:
```
var triggers = await _triggerStore.GetAsync();
if(triggers.Any() == false)
{
    await _triggerIndexer.IndexTriggersAsync();
}          
```  
But since IndexTriggersAsync is asynchronous/event driven I dont have any way of waiting, apart from maybe polling. 

Any ideas ?

**Comment by sfmskywalker** [comment]
Indeed, the indexing is implemented as an Elsa startup task, which means it is executed as part of the `StartupRunnerHostedService` hosted service.

If the Kafka message processer is implemented as a hosted service, you could register that service *after* doing `.AddElsa` (which internally registers the `StartupRunnerHostedService` hosted service).
